### PR TITLE
Deprecation: QAD aux wire

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** We want to deprecate passing `None` to `aux_wire` when using a mode that needs an `aux_wire` as we now use the resources available to choose a mode and not vice versa.

**Description of the Change:** Adds a deprecation for v0.45.

**Benefits:** Warn the users that the automatic finding of auxiliary wires will be removed in the next version.

**Possible Drawbacks:** Deprecates a feature that may have some (probably little) use.

**Related ShortCut Stories:** [sc-107729]

### CODE REVIEW

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -9,6 +9,12 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* Providing a value of None to aux_wire in reversed or standard mode has been deprecated and will
+  no longer be supported in 0.45. An aux_wire will no longer be automatically assigned.
```
Comment (astralcai) [inline doc/development/deprecations.rst]: Can you also update it to mention the function?

**Review by astralcai — APPROVED** [review]

File: pennylane/gradients/hadamard_gradient.py [hunk 2]
```diff
@@ -440,6 +440,14 @@ def hadamard_grad(
     # Validate or get default for aux_wire
     # unless using direct or reversed-direct modes
 
+    if mode in ["standard", "reversed"] and aux_wire is None:
+        raise PennyLaneDeprecationWarning(
+            """
+            Providing a value of None to aux_wire in reversed or standard mode has been deprecated and will 
+            no longer be supported in 0.46. An aux_wire will no longer be automatically assigned.
```
Comment (gabrielasd) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
            no longer be supported in v0.46. An aux_wire will no longer be automatically assigned.
```

**Review by gabrielasd — APPROVED** [review]
LGTM!

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -52,6 +52,10 @@
 
 <h3>Deprecations 👋</h3>
 
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` in reversed or standard mode has been
+  deprecated and will no longer be supported in 0.45. An ``aux_wire`` will no longer be automatically assigned.
```
Comment (andrijapau) [inline doc/releases/changelog-dev.md]:
```suggestion
  deprecated and will no longer be supported in 0.46. An ``aux_wire`` will no longer be automatically assigned.
```

File: doc/development/deprecations.rst [hunk 4]
```diff
@@ -9,6 +9,12 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` in reversed or standard mode has been
+  deprecated and will no longer be supported in 0.45. An ``aux_wire`` will no longer be automatically assigned.
```
Comment (andrijapau) [inline doc/development/deprecations.rst]:
```suggestion
  deprecated and will no longer be supported in 0.46. An ``aux_wire`` will no longer be automatically assigned.
```

File: tests/gradients/core/test_hadamard_gradient.py [hunk 5]
```diff
@@ -499,6 +499,19 @@ def test_automatic_mode_multiple_observables(self, mocker):
...
+
+        op = qml.evolve(qml.X(0) @ qml.X(1) + qml.Y(2) + qml.Z(0) @ qml.Z(1), t)
+        mp = qml.expval(qml.Z(0) @ qml.X(1) + qml.Y(0) + qml.X(0) @ qml.Z(1))
+        tape = qml.tape.QuantumScript([op], [mp, qml.probs((0, 1))])
+
+        with pytest.raises(
+            PennyLaneDeprecationWarning,
+            match="Providing a value of None to aux_wire",
+        ):
+            _, _ = qml.gradients.hadamard_grad(tape, mode="standard", aux_wire=None)
```
Comment (andrijapau) [inline tests/gradients/core/test_hadamard_gradient.py:514]: Don't we need to check if it's deprecated for "reversed" to?

**Review by andrijapau — COMMENTED** [review]
Initial set of comments, looking now at the test changes trying to understand them

File: tests/gradients/core/test_hadamard_gradient.py (hunk 5 above)
Comment (comp-phys-marc) [reply tests/gradients/core/test_hadamard_gradient.py:514]: I mean, there is just one place where we raise if the mode is standard or reversed and we don't have an aux_wire.

Comment (comp-phys-marc) [reply tests/gradients/core/test_hadamard_gradient.py:514]: But yeah I can test both cases.

Comment (comp-phys-marc) [reply tests/gradients/core/test_hadamard_gradient.py:514]: Done in https://github.com/PennyLaneAI/pennylane/pull/8905/commits/52dc7a58f038e8b541fe151213c317d6ac5ad29e

File: tests/workflow/interfaces/qnode/test_torch_qnode.py [hunk 6]
```diff
@@ -883,7 +919,13 @@ def circuit(x):
...
         x = torch.tensor([1.0, 2.0], requires_grad=True, dtype=torch.float64)
         res = circuit(x)
-        jac_fn = lambda x: jacobian(circuit, x, create_graph=True)
+
+        if diff_method == "hadamard":
+            # Cannot find a way to specify an aux_wire that is not already in use by the circuit!
+            with pytest.warns(PennyLaneDeprecationWarning, match="aux_wire"):
+                jac_fn = lambda x: jacobian(circuit, x, create_graph=True)
+        else:
+            jac_fn = lambda x: jacobian(circuit, x, create_graph=True)
```
Comment (andrijapau) [inline tests/workflow/interfaces/qnode/test_torch_qnode.py]: Do we have a plan for how to fix this?

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: I don't even fully understand why in these examples I cannot for the life of me find a wire that is not "in use by the circuit". No matter what I specify as the `aux_wire`, it's in use. When the circuit clearly does not involve it at all. I tried the whole alphabet.

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: It seems to happen mostly in tests that use the `jacobian` like this.

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: So in this case, it seems that if I provide an aux wire, the circuit is modified internally to return the probs on that wire too, meaning it can't be used as an auxiliary.

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: I guess when we use `jacobian`, we run things backward and when we do so we deal with an entirely different circuit that was initially specified??

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: See the circuit specified in the test, and the circuit that is actually seen by `hadamard_gradient`...

Screenshot 2026-01-16 at 1 40 05 PM
Screenshot 2026-01-16 at 1 40 20 PM

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: @albi3ro any ideas what is happening here?

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: @andrijapau do you have any ideas about this?

Comment (andrijapau) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: I can probably take a look tomorrow

Comment (albi3ro) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: Hessians = derivative of derivative.  The aux wire gets used for the first derivative, so there's no aux wire left for the derivative of the derivative.

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: I see, so this seems like a problem.

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: So I need a way to specify an `aux_wire` before each call to `jacobian`. Can I change the `gradient_kwargs` after the first one?

Comment (comp-phys-marc) [reply tests/workflow/interfaces/qnode/test_torch_qnode.py]: Seems like if I make the `jac_fn` a `QNode` with new `gradient_kwargs` after the fact, the results are really wonky and cause errors. :/

File: doc/development/deprecations.rst [hunk 7]
```diff
@@ -9,6 +9,13 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` in reversed or standard mode has been
```
Comment (andrijapau) [inline doc/development/deprecations.rst]: Maybe? It's a bit more clear what exactly the deprecated configuration is.
```suggestion
* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` with ``mode="reversed"` or `mode="standard"` has been
```

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -147,6 +147,13 @@
 
 <h3>Deprecations 👋</h3>
 
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` in reversed or standard mode has been
+  deprecated and will no longer be supported in 0.46. An ``aux_wire`` will no longer be automatically assigned.
+  [(#8905)](https://github.com/PennyLaneAI/pennylane/pull/8905)
+
+* The ``transform_program`` property of ``QNode`` has been renamed to ``compile_pipeline``.
+  The deprecated access through ``transform_program`` will be removed in PennyLane v0.46.
+  [(#8906)](https://github.com/PennyLaneAI/pennylane/pull/8906)
```
Comment (andrijapau) [inline doc/releases/changelog-dev.md]:
```suggestion
```

Another merge artifact? 🤔

File: pennylane/gradients/hadamard_gradient.py
Comment (andrijapau) [inline pennylane/gradients/hadamard_gradient.py:120]: Can you add a sphinx warning for the deprecation?

**Review by andrijapau — CHANGES_REQUESTED** [review]
Looking good, just some minor changes requested 😄

File: pennylane/gradients/hadamard_gradient.py [hunk 9]
```diff
@@ -134,6 +135,11 @@ def hadamard_grad(
         Executing this circuit will provide the Jacobian in the form of a tensor, a tuple, or a
         nested tuple depending upon the nesting structure of measurements in the original circuit.
 
+    .. warning::
+        Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` with ``mode="reversed"`
```
Comment (andrijapau) [inline pennylane/gradients/hadamard_gradient.py]: Can you move this to the top above `Args`

Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: ok

**Review by andrijapau — CHANGES_REQUESTED** [review]
Sorry ... but there's a formatting issue 
Image

**Review by andrijapau — APPROVED** [review]

File: pennylane/workflow/resolution.py [hunk 10]
```diff
@@ -186,6 +186,17 @@ def _resolve_mcm_config(
...
+    if initial_config.derivative_order > 1 and (
+        "aux_wire" in initial_config.gradient_keyword_arguments
+        or (
+            "mode" in initial_config.gradient_keyword_arguments
+            and initial_config.gradient_keyword_arguments["mode"] in ("standard", "reversed")
+        )
+    ):
+        raise ValueError(
+            "Higher order derivatives with hadamard gradients in standard and reversed modes are not possible. "
+            "Please use direct or reversed-direct mode to perform a higher order derivative with a hadamard gradient."
```
Comment (gabrielasd) [inline pennylane/workflow/resolution.py]: This error message looks good to me, my only suggestion would be to include the adverb `instead` in the last sentence (not sure if the end of the sentence or somewhere in the middle is the best position for it)

Comment (comp-phys-marc) [reply pennylane/workflow/resolution.py]: Added the word "instead"

File: doc/development/deprecations.rst [hunk 11]
```diff
@@ -9,6 +9,13 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` with ``mode="reversed"` or `mode="standard"` has been
```
Comment (gabrielasd) [inline doc/development/deprecations.rst]: One of these `mode=...` has an extra (or not enough backticks), the text looks blue

Comment (comp-phys-marc) [reply doc/development/deprecations.rst]: Fixed thanks

File: doc/releases/changelog-dev.md [hunk 12]
```diff
@@ -147,6 +147,10 @@
 
 <h3>Deprecations 👋</h3>
 
+* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` in reversed or standard mode has been
+  deprecated and will no longer be supported in 0.46. An ``aux_wire`` will no longer be automatically assigned.
```
Comment (gabrielasd) [inline doc/releases/changelog-dev.md]:
```suggestion
* Providing a value of ``None`` to ``aux_wire`` of ``qml.gradients.hadamard_grad`` with ``mode="reversed"`` or ``mode="standard"`` has been
  deprecated and will no longer be supported in 0.46. An ``aux_wire`` will no longer be automatically assigned.
```
